lib/selection.js

```diff
@@ -0,0 +1,202 @@
+  }
+
+  /**
+   * @public
+   */
+  isEmpty () {
+    return this._items.length === 0
+  }
+
+  /**
+   * @public
+   */
+  get length () {
+    return this._items.length
+  }
+
+  /**
+   * Takes an index and returns the item at that index
+   * @public
+   * @param {number} index
+   */
+  get (index) {
+    return this._items[index]
+  }
+
+  /**
+   * Takes an index and an item and replaces the item at the index with the new item
+   * @public
+   * @param {number} index
+   * @param {SelectionItem} item
+   */
+  set (index, item) {
+    this._items[index] = item
+  }
+
+  /**
+   * Takes two indexes and swaps the items at those indexes
+   * @public
+   * @param {number} i
+   * @param {number} j
+   */
+  swap (i, j) {
+    const temp = this._items[i]
+    this._items[i] = this._items[j]
+    this._items[j] = temp
+  }
+
+  /**
+   * @public
+   */
+  clear () {
+    this._items.length = 0
+  }
+
+  * [Symbol.iterator] () {
+    const self = this
+    for (let i = 0; i < this._items.length; i++) {
+      const item = this._items[i]
+      yield {
+        ...item,
+        remove () {
+          self._items.splice(i, 1)
+          i -= 1
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Examples:
+ * existing: 1-10 | 1-10
+ * new_item: 8-12 | 10-15
+ * @param {SelectionItem} newItem
+ * @param {SelectionItem} existing
+ * @returns {boolean} returns true if the new item's lower end is intersecting with the existing item
+ */
+export function isLowerIntersecting (newItem, existing) {
+  return (newItem.from <= existing.to) && (newItem.from > existing.from) && (newItem.to > existing.to)
+}
+
+/**
+ * Examples:
+ * existing: 20-25 | 20-25
+ * new_item: 15-22 | 15-20
+ * @param {SelectionItem} newItem
+ * @param {SelectionItem} existing
+ * @returns {boolean} returns true if the new item's upper end is intersecting with the existing item
+ */
+export function isUpperIntersecting (newItem, existing) {
+  return (newItem.to >= existing.from) && (newItem.to < existing.to) && (newItem.from < existing.from)
+}
+
+/**
+ * Examples:
+ * existing: 10-20 | 10-20 | 10-20
+ * new_item: 12-15 | 20-20 | 15-20
+ * @param {SelectionItem} newItem
+ * @param {SelectionItem} existing
+ * @returns {boolean} returns true if the new item is completely inside the existing item
+ */
+export function isInsideExisting (newItem, existing) {
+  const existingIntervalSize = existing.to - existing.from
+  const newItemIntervalSize = newItem.to - newItem.from
+  return newItem.from >= existing.from && newItem.to <= existing.to && (newItemIntervalSize < existingIntervalSize)
+}
+
+/**
+ * Examples:
+ * existing: 10-20 | 10-20 | 10-20
+ * new_item: 10-21 | 09-20 | 10-20
+ * @param {SelectionItem} newItem
+ * @param {SelectionItem} existing
+ * @returns {boolean} returns true if the new item is covering the existing item
+ */
+export function isCoveringExisting (newItem, existing) {
+  return newItem.from <= existing.from && newItem.to >= existing.to
+}
```

Comment:
you don't need return types on any of these as tslib will detect that this always returns bool, the returns comments you could move into the bottom of the top comment

Reply:
these params should also be `{from: number, to: number}` or `MinimalSelectionItem` if you want to make something like that, not SelectionItem

Reply:
The return types are not there necessarily there to provide the types, but rather to leave a comment on **when** this type returns true. As you can see they all have a _"returns true if the new items..."_ explanation after the `@returns {boolean}`. To me it makes sense to put these comments to the `@returns` section, but let me know if you'd still like me to change it.

Reply:
Also, I like the idea of having a smaller `MinimalSelectionItem` typedef, so thanks. I added that too.